sql: plan multiple table readers on the gateway if plan is local

yuzefovich opened the pull request:
This commit changes the physical planner to create multiple concurrent
TableReaders for a single TableReader spec when the plan is local if
- there is no required ordering
- the scan is deemed safe and beneficial to be parallelized (i.e. there
are neither hard nor soft limits and the scans is guaranteed to return at
most 10k rows)
- there are no mutations in the plan
- the plan is fully supported natively by the vectorized engine (i.e. we
don't need to wrap any row-execution processors into the vectorized flow)
- there is still quota available for additional goroutines for these
parallel TableReaders.

This is achieved by partitioning the spans to scan according to the
leaseholder boundaries and giving each partition a separate TableReader.
All the streams from those table readers are then merged by the parallel
unordered synchronizer. We limit the fanout for a single stage of
a local plan by 64 as well as we have a node-wide semaphore on the
number of additional goroutines needed for these concurrent processors
(determined by the private `sql.local_scans.concurrency_limit` cluster
setting which defaults to 1024; using 0 will disable the parallelization
entirely). This new cluster setting only controls how many goroutines
can be used for additional parallel TableReaders for local plans when it
is safe to parallelize the scans; using 0 value will make us go back to
the previous behavior of having a single TableReader in all cases.

This change seems quite beneficial in general but especially so in case
of the locality optimized search. There, this change allows us to
perform all lookups into the remote regions (in case of a local miss) in
parallel and return the result once a row is retrieved, without having
wait for each lookup to come back. This is possible because the parallel
unordered synchronizer in the vectorized engine will eagerly cancel the
work once the flow transitions to draining.

The requirement of not having any wrapped row-execution processors comes
from the fact that they eagerly transition into the draining state once
they have produced all the output rows, and this might occur to early
resulting in the parallel unordered synchronizers canceling work that
still needs to be performed.

The resulting plan in the vectorized engine is then as follows:
```
                    limitOp
                       |
        ---- serial unordered sync ----
        |                              |
local region scan      -----  parallel unordered sync -----
                       |               |                   |
                     remote regions scans (parallel TableReaders)
```

However, if the flow is row-based and we were to use the same physical
plan, it would actually cause all lookups (both local and all remote)
run in parallel without the ability to eagerly cancel the lookups once
the response is retrieved. (This is the case because we start each
processor eagerly in its own goroutine if two consecutive processors
cannot be fused together.) As a result, this change prohibits the
parallelization of scans in the local flows when the locality
optimized search is enabled and the vectorized mode is set to `off`
(the previous behavior is kept which is suboptimal in case of a local
miss).

Fixes: #68557.

Release note (performance improvement): If the query is executed locally
and needs to perform reads from multiple remote leaseholders, those
remote reads might now be done faster. This is especially likely for the
case of locality optimized search when there is a local region miss.

Release justification: low-risk improvement to the existing
functionality.

yuzefovich committed:
sql: plan multiple table readers on the gateway if plan is local

This commit changes the physical planner to create multiple concurrent
TableReaders for a single TableReader spec when the plan is local if
- there is no required ordering
- the scan is deemed safe and beneficial to be parallelized (i.e. there
are neither hard nor soft limits and the scans is guaranteed to return at
most 10k rows)
- there are no mutations in the plan
- the plan is fully supported natively by the vectorized engine (i.e. we
don't need to wrap any row-execution processors into the vectorized flow)
- there is still quota available for additional goroutines for these
parallel TableReaders.

This is achieved by partitioning the spans to scan according to the
leaseholder boundaries and giving each partition a separate TableReader.
All the streams from those table readers are then merged by the parallel
unordered synchronizer. We limit the fanout for a single stage of
a local plan by 64 as well as we have a node-wide semaphore on the
number of additional goroutines needed for these concurrent processors
(determined by the private `sql.local_scans.concurrency_limit` cluster
setting which defaults to 1024; using 0 will disable the parallelization
entirely). This new cluster setting only controls how many goroutines
can be used for additional parallel TableReaders for local plans when it
is safe to parallelize the scans; using 0 value will make us go back to
the previous behavior of having a single TableReader in all cases.

This change seems quite beneficial in general but especially so in case
of the locality optimized search. There, this change allows us to
perform all lookups into the remote regions (in case of a local miss) in
parallel and return the result once a row is retrieved, without having
wait for each lookup to come back. This is possible because the parallel
unordered synchronizer in the vectorized engine will eagerly cancel the
work once the flow transitions to draining.

The requirement of not having any wrapped row-execution processors comes
from the fact that they eagerly transition into the draining state once
they have produced all the output rows, and this might occur to early
resulting in the parallel unordered synchronizers canceling work that
still needs to be performed.

The resulting plan in the vectorized engine is then as follows:
```
                    limitOp
                       |
        ---- serial unordered sync ----
        |                              |
local region scan      -----  parallel unordered sync -----
                       |               |                   |
                     remote regions scans (parallel TableReaders)
```

However, if the flow is row-based and we were to use the same physical
plan, it would actually cause all lookups (both local and all remote)
run in parallel without the ability to eagerly cancel the lookups once
the response is retrieved. (This is the case because we start each
processor eagerly in its own goroutine if two consecutive processors
cannot be fused together.) As a result, this change prohibits the
parallelization of scans in the local flows when the locality
optimized search is enabled and the vectorized mode is set to `off`
(the previous behavior is kept which is suboptimal in case of a local
miss).

Release note (performance improvement): If the query is executed locally
and needs to perform reads from multiple remote leaseholders, those
remote reads might now be done faster. This is especially likely for the
case of locality optimized search when there is a local region miss.

Release justification: low-risk improvement to the existing
functionality.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/68679)
<!-- Reviewable:end -->

yuzefovich committed:
sql: force serialization explicitly for locality optimized search

Previously, we were relying on some implicit things (namely on the usage
of `opt.FuseAggressively` during the local flow setup) to serialize the
two arms of the locality optimized search. This is no longer needed
after we introduced an explicit SERIAL_UNORDERED synchronizer type
a couple of months ago.

Note that there are a couple of changes to the physical plans in the
tests, but they are only applicable for EXPLAIN runs of the query since
during the actual query run previously we would still use the serial
unordered sync correctly.

Release note: None

Release justification: low-risk pure cleanup.

yuzefovich commented:
I think this is RFAL.

Curious to hear everyone's thought on these questions:
- do we want to remove the setting that enables this planning altogether (i.e. have it enabled all the time)? To me it seems like it's always beneficial to parallelize if the optimizer deemed it "safe and beneficial".
- do we want to limit the concurrency here? If so, should the limit be global (similar to 1024 async partial batch requests at the DistSender level)?

yuzefovich commented:
Just did a quick sanity check that it works as expected. I created 9 node cluster with us-west, us-central, and europe-west regions, then inserted a row with us-central as its home and tried to query it from us-west. Before this PR the latency was 140ms (RTT from us-west to europe-west), after this PR the latency is 50ms (RTT from us-west to us-central). One slightly annoying thing is that the context cancellation shows up in the trace, yet it doesn't affect the query execution.

Here are the bundles for the curious: [slow](https://github.com/cockroachdb/cockroach/files/6994903/stmt-bundle-685084668180496385.zip) and [fast](https://github.com/cockroachdb/cockroach/files/6994905/stmt-bundle-685084889233457153.zip).

rytaft reviewed (commented):
Wow that's great! Really exciting to hear that this will reduce latency so significantly.

> do we want to remove the setting that enables this planning altogether (i.e. have it enabled all the time)? To me it seems like it's always beneficial to parallelize if the optimizer deemed it "safe and beneficial".

Is the optimizer actually impacting this decision right now? Seems like this logic applies whether or not the optimizer has chosen to plan locality optimized search.

But anyway, I agree that it seems like this parallelization should always be beneficial, but it also might be useful to be able to switch this off if we find that it's not behaving as expected (e.g., the concurrency turns out to be too high). I agree with your decision to make it private so users don't try to mess with it, but it could be a useful switch for us to have available when providing customer support.

> do we want to limit the concurrency here? If so, should the limit be global (similar to 1024 async partial batch requests at the DistSender level)?

I agree it's probably good to have some limit on the concurrency.  I don't have a good sense of what the limit should be, though...

> Reviewed 6 of 6 files at r1, 2 of 2 files at r2, 14 of 14 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-:-MhFy5l06WNa_oR2oM2Z:bssyey9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2, @RaduBerinde, and @yuzefovich)

___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row, line 832 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-MhFngW42IE9DkKpf7W_:-MhFngW56oy6KAtfnwlO:b8cjvuc) ([raw file](https://github.com/cockroachdb/cockroach/blob/f30d963878073fc9a5386ebd47ae7eabf1525a8d/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row#L832)):*
> ```
> ----
> Scan /Table/73/1/"\xc0"/1/0
> Scan /Table/73/1/"@"/1/0, /Table/73/1/"\x80"/1/0
> ```

Is there a way to tell from this trace that these scans are all happening in parallel? Might be worth updating the comment to mention what to look for.

Also, did you mean to put two of the leaseholders on the same node? Seems like you might need to update the splits above so we get 3 scans instead of two (if that was your intention).
___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row, line 862 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-MhFoL3o8wD0JDGfsrLI:-MhFoL3pEaNCq4RGwOmB:b8gnt1g) ([raw file](https://github.com/cockroachdb/cockroach/blob/f30d963878073fc9a5386ebd47ae7eabf1525a8d/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row#L862)):*
> ```
>       spans: [/'ca-central-1'/1 - /'ca-central-1'/1] [/'us-east-1'/1 - /'us-east-1'/1]
> ·
> Diagram: https://cockroachdb.github.io/distsqlplan/decode.html#eJysklFr2zAQx9_3KcQ9bUMjlpPCEAw8NocFsqSzDRvUJqjWEcQcy5Nk1hL83Yflh84jC43TR53up_vd3z6C_VUBhzRex58y8pYsk-1Xchf_uF1_XG3I68-rNEu_rd-QcYPBvdK1qHb3jzujf--cuK-QfP8SJzFpfpIPhBVku1ymcUZCoFBriRtxQAv8DhgUFBqjS7RWm7509A0r-QA8oKDqpnV9uaBQaoPAj-CUqxA4ZP2cBIVEMwuAgkQnVOWfPakUNUYdhHkECmkjasvJLIc8f3gf5DBj_olt6ziJGI1CGs1ptKDRDRQdBd26Jw_rxB6Bs45Oc2VXuJYv7fqk2NbaSDQoR3pFd2KbjX6nm1n4T-Pp0eHUmOZTY4omRTR_RkQWjRIVuTipxXiVtTooR9h_VYJLIkvQNrq2-KyPEfSOKPc47GV1a0q8Nbr0Y4bj1nO-ING64TYcDqvaX_nf6W-YXQOHZ-HFeXh-Dbw4C9-M4KAruld_AgAA__9L0Kp2
> ```

Not for this PR, but it would be great to show in the DistSQL diagram which of the two inputs is executed first in the serial unordered case.
___
*[pkg/sql/distsql_physical_planner.go, line 668 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-MhFqHBe08plfmyTM7ev:-MhFqHBe08plfmyTM7ew:b-lp5zwx) ([raw file](https://github.com/cockroachdb/cockroach/blob/f30d963878073fc9a5386ebd47ae7eabf1525a8d/pkg/sql/distsql_physical_planner.go#L668)):*
> ```Go
> 	// want to capture that value in case there is a race between actual
> 	// physical planning and updating the setting on a separate connection) and
> 	// on the fact whether there are any mutations in the plan (which prohibit
> ```

nit: on the fact whether -> whether
___
*[pkg/sql/distsql_physical_planner.go, line 3800 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-MhFs_eYAq8xoBpJC6Wx:-MhFs9E5EQwbPKpZChPz:bibogtd) ([raw file](https://github.com/cockroachdb/cockroach/blob/f30d963878073fc9a5386ebd47ae7eabf1525a8d/pkg/sql/distsql_physical_planner.go#L3800)):*
> ```Go
> 	if !distribute {
> 		if planner == nil || dsp.spanResolver == nil || planner.curPlan.flags.IsSet(planFlagContainsMutation) {
> 			planCtx.parallelizeScansIfLocal = false
> ```

nit: might help to add a comment here to explain why we can't parallelize in this case
___
*[pkg/sql/distsql_running.go, line 441 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-MhFtqCjE49C7rVqDaPJ:-MhFtqCjE49C7rVqDaPK:b17q736) ([raw file](https://github.com/cockroachdb/cockroach/blob/f30d963878073fc9a5386ebd47ae7eabf1525a8d/pkg/sql/distsql_running.go#L441)):*
> ```Go
> 		}
> 	}
> 	if (!localState.IsLocal || localState.HasConcurrency) && txn != nil {
> ```

nit: feels like you're checking these two bools in a lot of places now. Maybe it's worth making a helper function that checks both?
___
*[pkg/sql/distsql_running.go, line 526 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-MhFtHIb93eGjsET4KtU:-MhFtHIb93eGjsET4KtV:bubh4wf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f30d963878073fc9a5386ebd47ae7eabf1525a8d/pkg/sql/distsql_running.go#L526)):*
> ```Go
> 	// This is important, since these flows are forced to use the RootTxn (since
> 	// they might have mutations), and the RootTxn does not permit concurrency.
> 	// For such flows, we were supposed to have fused everything.
> ```

nit: comment needs a tweak
___
*[pkg/sql/logictest/testdata/logic_test/vectorize_local, line 226 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-MhFuMPwAXlDxVS79CRM:-MhFuMPwAXlDxVS79CRN:b-ismxv2) ([raw file](https://github.com/cockroachdb/cockroach/blob/f30d963878073fc9a5386ebd47ae7eabf1525a8d/pkg/sql/logictest/testdata/logic_test/vectorize_local#L226)):*
> ```
> querying next range at /Table/56/1/0/0
> querying next range at /Table/56/1/10/0
> querying next range at /Table/56/1/0/0
> ```

why are these same ranges getting queried twice?
___
*[pkg/sql/opt/exec/execbuilder/testdata/select, line 1779 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-MhFupa21urHuyywj-ym:-MhFupa21urHuyywj-yn:b-5rbbq0) ([raw file](https://github.com/cockroachdb/cockroach/blob/f30d963878073fc9a5386ebd47ae7eabf1525a8d/pkg/sql/opt/exec/execbuilder/testdata/select#L1779)):*
> ```
> querying next range at /Table/74/1/0/0
> querying next range at /Table/74/1/10/0
> querying next range at /Table/74/1/0/0
> ```

another duplicate range scan?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Is the optimizer actually impacting this decision right now? Seems like this logic applies whether or not the optimizer has chosen to plan locality optimized search.

In `planTableReaders` in this new code block
```
		if len(info.reqOrdering) == 0 &&
			info.parallelize &&
			planCtx.parallelizeScansIfLocal &&
			!prohibitParallelScans &&
			planCtx.spanIter != nil {
```
`info.parallelize` comes from `scanNode.parallelize` which is set by the optimizer here: https://github.com/cockroachdb/cockroach/blob/4aa67c383993d2ef0173b181e8f147ab316b7dfd/pkg/sql/opt/exec/execbuilder/relational.go#L598-L610

----

I think I'll introduce a global semaphore to limit the concurrency with the limit controlled by a cluster setting. The default value will be 1024, and if set to 0, this will effectively disable the parallelization of local scans at the TableReader level. This way we will still have a knob to disable this feature if necessary while having only a single setting. Thoughts?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-:-MhJmxgtB47RUzy8qWUa:bua9e0x)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2, @RaduBerinde, and @rytaft)

___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row, line 832 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-MhFngW42IE9DkKpf7W_:-MhJkcQl4qJ-4UXh6EG0:b-eiondy) ([raw file](https://github.com/cockroachdb/cockroach/blob/f30d963878073fc9a5386ebd47ae7eabf1525a8d/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row#L832)):*
> Is there a way to tell from this trace that these scans are all happening in parallel?

I don't think it's possible - I think we have to look at the DistSQL diagram of `EXPLAIN (VEC)` to see that we have two TableReaders with a parallel unordered synchronizer on top of them meaning that both processors run concurrently.

I extended the comment.

> Also, did you mean to put two of the leaseholders on the same node? Seems like you might need to update the splits above so we get 3 scans instead of two (if that was your intention).

I did set up 3 leaseholders, and I'm guessing that having two scans here is expected because even if the locality optimized search is disabled, we still divide all spans to scans into "local" and "remote" in the optimizer, right? The difference is that we don't set a hard limit on the UnionAll, and during the physical planning that is converted into a parallel unordered synchronizer.
___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row, line 862 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-MhFoL3o8wD0JDGfsrLI:-MhJdob7-p6vG3Vxhra1:b-bpzb9j) ([raw file](https://github.com/cockroachdb/cockroach/blob/f30d963878073fc9a5386ebd47ae7eabf1525a8d/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row#L862)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Not for this PR, but it would be great to show in the DistSQL diagram which of the two inputs is executed first in the serial unordered case.
</blockquote></details>

I think we can already see this - it'll always be the left input once the diagram is spread out
![Screen Shot 2021-08-17 at 8.08.03 AM.png](https://files.reviewableusercontent.io/images/reviewable/5135385/lCd9D0-slfEuxxu5v-KxbQVh/Screen%20Shot%202021-08-17%20at%208.08.03%20AM.png)
___
*[pkg/sql/distsql_running.go, line 441 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-MhFtqCjE49C7rVqDaPJ:-MhJgaTd0RTl49Rp3M1Z:b-8rww7z) ([raw file](https://github.com/cockroachdb/cockroach/blob/f30d963878073fc9a5386ebd47ae7eabf1525a8d/pkg/sql/distsql_running.go#L441)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: feels like you're checking these two bools in a lot of places now. Maybe it's worth making a helper function that checks both?
</blockquote></details>

Added a helper, let me know what you think.
___
*[pkg/sql/logictest/testdata/logic_test/vectorize_local, line 226 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-MhFuMPwAXlDxVS79CRM:-MhJhBZG7fymOspnkTzT:b-pwc8rj) ([raw file](https://github.com/cockroachdb/cockroach/blob/f30d963878073fc9a5386ebd47ae7eabf1525a8d/pkg/sql/logictest/testdata/logic_test/vectorize_local#L226)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

why are these same ranges getting queried twice?
</blockquote></details>

This is because this PR adds the call to `PartitionSpans` to divide up the spans to scan during the physical planning (this internally uses the range cache of the DistSender); after the partitioning is performed, we will have two parallel TableReaders that also will go through the DistSender in order to perform the actual scans.

I updated the comment.
___
*[pkg/sql/opt/exec/execbuilder/testdata/select, line 1779 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-MhFupa21urHuyywj-ym:-MhJhrx14JVwrb8TYd6l:b-1fjwzh) ([raw file](https://github.com/cockroachdb/cockroach/blob/f30d963878073fc9a5386ebd47ae7eabf1525a8d/pkg/sql/opt/exec/execbuilder/testdata/select#L1779)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

another duplicate range scan?
</blockquote></details>

Same explanation as above.


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> info.parallelize comes from scanNode.parallelize which is set by the optimizer here:

Interesting -- I guess this means we must be setting parallelize to true for the remote scan in locality optimized search? Technically `maxResults` should be 1 for locality optimized search, but maybe `indexConstraintMaxResults` doesn't know that.

> I think I'll introduce a global semaphore...

SGTM

> Reviewed 19 of 19 files at r4, 19 of 19 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-:-MhK9vvnBZJak-EtzAtO:bd38ae1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2, @RaduBerinde, and @yuzefovich)

___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row, line 832 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-MhFngW42IE9DkKpf7W_:-MhKAmxq9YJCsaGroB7f:bmd3bqr) ([raw file](https://github.com/cockroachdb/cockroach/blob/f30d963878073fc9a5386ebd47ae7eabf1525a8d/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row#L832)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

> Is there a way to tell from this trace that these scans are all happening in parallel?

I don't think it's possible - I think we have to look at the DistSQL diagram of `EXPLAIN (VEC)` to see that we have two TableReaders with a parallel unordered synchronizer on top of them meaning that both processors run concurrently.

I extended the comment.

> Also, did you mean to put two of the leaseholders on the same node? Seems like you might need to update the splits above so we get 3 scans instead of two (if that was your intention).

I did set up 3 leaseholders, and I'm guessing that having two scans here is expected because even if the locality optimized search is disabled, we still divide all spans to scans into "local" and "remote" in the optimizer, right? The difference is that we don't set a hard limit on the UnionAll, and during the physical planning that is converted into a parallel unordered synchronizer.
</blockquote></details>

I think what's happening is `/Table/73/1/"\xc0"/1/0` is located on node 4, while `/Table/73/1/"@"/1/0` and `/Table/73/1/"\x80"/1/0` are both on node 1. If you want them all to be on different nodes, you should probably change the split statement on line 780 to something like:

```
statement ok
ALTER TABLE regional_by_row_table SPLIT AT VALUES ('ca-central-1', 0), ('us-east-1', 0);
ALTER TABLE regional_by_row_table EXPERIMENTAL_RELOCATE VALUES (ARRAY[1], 'ap-southeast-2', 0), (ARRAY[4], 'ca-central-1', 0), (ARRAY[7], 'us-east-1', 0);
```
___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row, line 862 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-MhFoL3o8wD0JDGfsrLI:-MhKEFQj0RzyBEY_sHzS:b42ol9t) ([raw file](https://github.com/cockroachdb/cockroach/blob/f30d963878073fc9a5386ebd47ae7eabf1525a8d/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row#L862)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I think we can already see this - it'll always be the left input once the diagram is spread out
![Screen Shot 2021-08-17 at 8.08.03 AM.png](https://files.reviewableusercontent.io/images/reviewable/5135385/lCd9D0-slfEuxxu5v-KxbQVh/Screen%20Shot%202021-08-17%20at%208.08.03%20AM.png)

</blockquote></details>

Hmm but it's easy to rearrange in the diagram, and I don't think it's obvious which way to spread them out:
![Screen Shot 2021-08-17 at 12.51.01 PM.png](https://files.reviewableusercontent.io/images/reviewable/3288732/QSOw4jDsiId8ZvS1dOKvdv3J/Screen%20Shot%202021-08-17%20at%2012.51.01%20PM.png)
___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row, line 827 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-MhKBBJME5KIS8aou5ZU:-MhKBBJNDwHLxIjleO87:b-st1f) ([raw file](https://github.com/cockroachdb/cockroach/blob/845d4926eb8aacedb9923df4eef2838443e0e755/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row#L827)):*
> ```
> # to see from the trace that the lookups into the local and remote regions
> # happen in parallel: we can only see that all spans are divided into "local"
> # and "remote" (and the relevant "Scan" tracing messages can be in any order).
> ```

I don't think this is accurate, since I think "@" is the local region. I think the fact that you have two rows instead of 3 is just an accident of how you defined the splits
___
*[pkg/sql/distsql_running.go, line 441 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-MhFtqCjE49C7rVqDaPJ:-MhKEmXI3S02wLXw8_W0:bb6ywmx) ([raw file](https://github.com/cockroachdb/cockroach/blob/f30d963878073fc9a5386ebd47ae7eabf1525a8d/pkg/sql/distsql_running.go#L441)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Added a helper, let me know what you think.
</blockquote></details>

LGTM


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Added the semaphore with 1024 default limit.

> > info.parallelize comes from scanNode.parallelize which is set by the optimizer here:
> 
> Interesting -- I guess this means we must be setting parallelize to true for the remote scan in locality optimized search? Technically maxResults should be 1 for locality optimized search, but maybe indexConstraintMaxResults doesn't know that.

Yeah, I was surprised by this behavior initially too (thinking that the limit should be set which will leave `parallelize` unchanged); however, I have now come to appreciate this probably-bug. I think it's a good idea stability-wise to parallelize the scans only if we're guaranteed to retrieve not many rows (so it's nice to require optimizer-driven `parallelize` be `true`), and also the locality optimized search satisfies that condition. I'm thinking about leaving everything as is and not fixing this probably-bug in 21.2 timeframe unless we see some negative effects because of it. Thoughts?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-:-MhLVekz8AiENw-o1qZH:b-j5zeci)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2, @RaduBerinde, and @rytaft)

___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row, line 832 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-MhFngW42IE9DkKpf7W_:-MhLYbs_Btj4SCLjYQsl:b-45j4ap) ([raw file](https://github.com/cockroachdb/cockroach/blob/f30d963878073fc9a5386ebd47ae7eabf1525a8d/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row#L832)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I think what's happening is `/Table/73/1/"\xc0"/1/0` is located on node 4, while `/Table/73/1/"@"/1/0` and `/Table/73/1/"\x80"/1/0` are both on node 1. If you want them all to be on different nodes, you should probably change the split statement on line 780 to something like:

```
statement ok
ALTER TABLE regional_by_row_table SPLIT AT VALUES ('ca-central-1', 0), ('us-east-1', 0);
ALTER TABLE regional_by_row_table EXPERIMENTAL_RELOCATE VALUES (ARRAY[1], 'ap-southeast-2', 0), (ARRAY[4], 'ca-central-1', 0), (ARRAY[7], 'us-east-1', 0);
```
</blockquote></details>

Oh, indeed, good catch. This makes a lot more sense, thanks!
___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row, line 862 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-MhFoL3o8wD0JDGfsrLI:-MhLZ7Sk2ZypEfeZSyHn:b-rdxnoq) ([raw file](https://github.com/cockroachdb/cockroach/blob/f30d963878073fc9a5386ebd47ae7eabf1525a8d/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row#L862)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Hmm but it's easy to rearrange in the diagram, and I don't think it's obvious which way to spread them out:
![Screen Shot 2021-08-17 at 12.51.01 PM.png](https://files.reviewableusercontent.io/images/reviewable/3288732/QSOw4jDsiId8ZvS1dOKvdv3J/Screen%20Shot%202021-08-17%20at%2012.51.01%20PM.png)

</blockquote></details>

Huh, I didn't know that it was possible to rearrange the order of inputs - I thought they were fixed. I agree with you that it'd be good to show which arm is executed first.
___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row, line 827 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-MhKBBJME5KIS8aou5ZU:-MhLYfPkE6wVYYeZdkEk:bqxqqbn) ([raw file](https://github.com/cockroachdb/cockroach/blob/845d4926eb8aacedb9923df4eef2838443e0e755/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row#L827)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I don't think this is accurate, since I think "@" is the local region. I think the fact that you have two rows instead of 3 is just an accident of how you defined the splits
</blockquote></details>

Indeed, fixed.

Now I had to add some `rowsort` options (since there is a race between "remote" regions) as well as have separate tracing paths for vec vs non-vec (for the former we have two parallel TableReader, and for the latter we have a single TableReader with two spans).


<!-- Sent from Reviewable.io -->

yuzefovich commented:
The CI failures [here](https://teamcity.cockroachdb.com/viewLog.html?buildId=3316656&buildTypeId=Cockroach_UnitTests)
```
        testdata/logic_test/regional_by_row:1458: 
        expected success, but found
        (XXUUU) txn already encountered an error; cannot be used anymore (previous err: aborted during DistSender.Send: context canceled)
```
worry me a bit - maybe the current approach of using the context cancellation to interrupt no longer needed scan requests "poisons" the txn, and, thus, we might have a correctness issue.

@nvanbenschoten @andreimatei I have a question for you: if we have an explicit transaction going on, and during a query evaluation we cancel some of the issued KV batch requests via the context cancellation, does the txn become poisoned and is no longer usable? We have a fully local query plan and are using the LeafTxn if that matters.

----

However, that still wouldn't explain the failures in the CI - there, it seems like we have a race between fully completing one parallel TableReader (that finds the row being looked up) and just setting up another one. At the moment, I'm not sure how to go around that or make it non-flaky.

andreimatei commented:
> @nvanbenschoten @andreimatei I have a question for you: if we have an explicit transaction going on, and during a query evaluation we cancel some of the issued KV batch requests via the context cancellation, does the txn become poisoned and is no longer usable? We have a fully local query plan and are using the LeafTxn if that matters.

Any non-retriable error in a root txn and any error in a leaf txn make the respective txn unusable. There are some explicitly-listed errors that allow a transaction to recover by rolling back to a savepoint:
https://github.com/cockroachdb/cockroach/blob/b761eba20b218aee12b295df027d22c4a17c9b61/pkg/kv/kvclient/kvcoord/txn_coord_sender.go#L875

In the write case, recovering from an arbitrary error (in particular, a communication error or a cancellation) is hard because we don't know the maximum timestamp at which we wrote - and that info is important for dealing with the intents we might have left behind one way or another. In the read case, we could be more lax if need be. Different leafs are already independent of one another.

Does this help?

rytaft reviewed (commented):
> Reviewed 14 of 14 files at r6, 15 of 19 files at r7, 1 of 1 files at r9, 1 of 1 files at r10, 1 of 1 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68679)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2, @RaduBerinde, @rytaft, and @yuzefovich)

___
*[pkg/sql/distsql_physical_planner.go, line 1410 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-MhPnDsnDoJBgqqxlSu7:-MhPnDsnDoJBgqqxlSu8:b-oup4b8) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa2e824dae305ce536e597f5cd28fe534eea781/pkg/sql/distsql_physical_planner.go#L1410)):*
> ```Go
> 			//
> 			// Determine the desired concurrency.
> 			const maxConcurrency = 64
> ```

Should this also be a setting?
___
*[pkg/sql/distsql_physical_planner.go, line 1422 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-MhPnmpl0vTB9aPaA7KB:-MhPnmpl0vTB9aPaA7KC:b-ntyxp4) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa2e824dae305ce536e597f5cd28fe534eea781/pkg/sql/distsql_physical_planner.go#L1422)):*
> ```Go
> 					break
> 				}
> 				actualConcurrency--
> ```

why are you decrementing here?


<!-- Sent from Reviewable.io -->

yuzefovich commented:
@andreimatei Some additional info: we will trigger the context cancellation error only when we are draining the local flow (this will occur only for fully local physical plans that don't have any writes/mutations), and the LeafTxn will not be used for that statement anymore (other than collecting the LeafTxnFinalState which works ok). I want to consider the following scenario: we have an explicit transaction consisting of multiple statements, for the first statement we create a LeafTxn that will be in error state once we drain the flow, will the explicit RootTxn become poisoned too? Will consequent statement be affected in any way? I guess I might as well just try this scenario and since what happens, but it'd be good to get your understanding too.

andreimatei commented:
There's no telepathic communication between a leaf and the root, so I think what you're describing is supposed to work fine. A leaf's error can be disregarded. You can read around here to see what information travels from a leaf to a root:
https://github.com/cockroachdb/cockroach/blob/b761eba20b218aee12b295df027d22c4a17c9b61/pkg/kv/kvclient/kvcoord/txn_coord_sender.go#L1128
This `LeafTxnFinalState` guy travels as metadata through the flow and is ingested into the root. I think there's no trace of an error in it.

rytaft commented:
> I'm thinking about leaving everything as is and not fixing this probably-bug in 21.2 timeframe unless we see some negative effects because of it. Thoughts?

SGTM





<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> There's no telepathic communication between a leaf and the root, so I think what you're describing is supposed to work fine. A leaf's error can be disregarded.

Great, thanks, we're good then.

----

I had to relax `scan_parallel` test a bit (for some reason we non-deterministically see Get and Scan requests when only Get requests should occur, I'm not sure why, but I don't think this is a blocking question); however, I still don't understand why we're seeing "txn already encountered an error" in `regional_by_row` file when actually running the query, looking more into this.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-:-MhQ3AyQBG8S7DTzlV0T:bk2qghc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2, @RaduBerinde, and @rytaft)

___
*[pkg/sql/distsql_physical_planner.go, line 1410 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-MhPnDsnDoJBgqqxlSu7:-MhQ3JGdAvEd8XL79S7A:b-edz4fe) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa2e824dae305ce536e597f5cd28fe534eea781/pkg/sql/distsql_physical_planner.go#L1410)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Should this also be a setting?
</blockquote></details>

I feel like no - I doubt we'll need to change it.

I'm a bit on the fence on whether we want to have this fanout limit though - it'll matter only for large clusters if the scan to parallelize touches most nodes. For example, in 256 node cluster if we need to lookup into ranges located on each of the nodes, `len(spanPartitions)` will be 256; with 1024 default concurrency 4 such queries will use up the quota; this limit of 64 adds a bit more "fairness" while possibly slowing down the parallel lookups. I guess there might be some other benefits to having less parallelism once a certain threshold is exceeded. WDYT?
___
*[pkg/sql/distsql_physical_planner.go, line 1422 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-MhPnmpl0vTB9aPaA7KB:-MhQ3yrO5L0b0QNSoKh0:b-rwzwvx) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa2e824dae305ce536e597f5cd28fe534eea781/pkg/sql/distsql_physical_planner.go#L1422)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

why are you decrementing here?
</blockquote></details>

We start with maximum number trying to acquire the goroutine quota, if that doesn't succeed, we decrement by one and try the quota again. Extended the comment.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
Oh, I think I see the problem with the existing approach. Consider this vectorized plan:
```
  └ *rowexec.joinReader
   └ *colexec.Materializer
    └ *colexec.limitOp
      └ *colexec.SerialUnorderedSynchronizer
        ├ *colfetcher.ColBatchScan
        └ *colexec.ParallelUnorderedSynchronizer
          ├ *colfetcher.ColBatchScan
          └ *colfetcher.ColBatchScan
```
If we have a local region miss, then we will run the parallel unordered sync arm which will return a batch with 1 row, and eventually it'll make it to the joinReader. The joinReader will then ask for another row from the materializer, which will ask the limitOp which will return a zero-length batch. At this point, the materializer will move into the draining state which will prompt the parallel unordered sync to cancel its input contexts which will might cancel the currently-in-flight lookup into another remote region, thus, poisoning the txn. What's important is that the flow as a whole has not finished the execution yet - we are still buffering the input rows in the joinReader and haven't performed the lookup there, and once we try to do it, we get the "txn already encountered an error" issue.

Possibly we might have to limit the solution only to the cases of the vectorized flows without any wrapped processors.

yuzefovich reviewed (commented):
I added another couple of commits that performs the walk over the plan to see whether it can be fully supported by the native vectorized operators, so we don't run into a scenario as described above. This is a bit ugly but should work correctly in all settings. (The second WIP commit is a minor extension which I'm not sure is worth it.)

An alternative solution would be to optimistically assume that it'll be safe to parallelize, then setup parallel TableReaders during the physical planning, and in case we need to plan wrapped row-by-row processors, squash the parallel TableReaders into a single one during the vectorized flow setup. This might look a bit better, but the EXPLAIN variants would be incorrect.

Overall, I'm a bit sad that this PR has gotten so ugly and limited. It still accomplishes the fast lookups into the nearby regions in limited scenarios, but it adds quite a bit of the complexity. I'm leaning towards proceeding with it anyway but curious to hear thoughts of other folks.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-:-MhUWt-ICEjKU9ZIa2bC:b-4e0vfi)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2, @RaduBerinde, and @rytaft)



<!-- Sent from Reviewable.io -->

rytaft commented:
So the optimization won't apply if there are any lookup/index joins in the plan?

Maybe we should check and see if this will satisfy the desired customer use cases.





<!-- Sent from Reviewable.io -->

yuzefovich commented:
> So the optimization won't apply if there are any lookup/index joins in the plan?

Yeah. Given Drew's recent work we can extend the use case to include the index joins (we'll need to be careful to confirm that we're not touching the interleaved indexes), but lookup joins will prohibit the parallelization of scans.

yuzefovich commented:
I included the index join on non-interleaved tables.

Our customer success team has a meeting scheduled with the customer on Tuesday, and they will confirm whether this proposed solution is sufficient (at least in some cases). I don't think we have time to implement any other approach for 21.2 anyway.

rytaft reviewed (commented):
Sounds good.

> Reviewed 16 of 17 files at r17, 17 of 17 files at r18, 1 of 2 files at r19, 1 of 1 files at r21, 1 of 1 files at r22, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-:-MhnZaVx3I-69E9irUP2:brxp9sq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2, @RaduBerinde, @rytaft, and @yuzefovich)

___
*[-- commits, line 36 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-MhnVnV_5cw2o03ClQEk:-MhnVnVa4O66YfoRtfC5:bak142w) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3dc37a73851322efe7c102e4f66974a52c3c41a/-- commits#L36)):*
What else is this setting used for? Will setting it to 0 have any other unintended effects?
___
*[pkg/sql/distsql_physical_planner.go, line 1410 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-MhPnDsnDoJBgqqxlSu7:-MhnWOje-b2JAjsvnB8V:b-c7mbc0) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa2e824dae305ce536e597f5cd28fe534eea781/pkg/sql/distsql_physical_planner.go#L1410)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I feel like no - I doubt we'll need to change it.

I'm a bit on the fence on whether we want to have this fanout limit though - it'll matter only for large clusters if the scan to parallelize touches most nodes. For example, in 256 node cluster if we need to lookup into ranges located on each of the nodes, `len(spanPartitions)` will be 256; with 1024 default concurrency 4 such queries will use up the quota; this limit of 64 adds a bit more "fairness" while possibly slowing down the parallel lookups. I guess there might be some other benefits to having less parallelism once a certain threshold is exceeded. WDYT?
</blockquote></details>

I think having a limit is a good idea. Seems useful to prevent a small number of queries from saturating all resources.
___
*[pkg/sql/distsql_physical_planner.go, line 3894 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-MhnZlxH26Re375LSrkM:-MhnZlxH26Re375LSrkN:b49x9cy) ([raw file](https://github.com/cockroachdb/cockroach/blob/6963da28b440c20d42bff997917957ea5f628407/pkg/sql/distsql_physical_planner.go#L3894)):*
> ```Go
> }
> 
> // canParallelizeScansIfLocal returns whether the plan contains scanNodes that
> ```

nit: canParallelizeScansIfLocal -> checkScanParalleizationIfLocal
___
*[pkg/sql/distsql_physical_planner.go, line 3905 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-MhnZseJ7K2CsX8lv1GR:-MhnZseJ7K2CsX8lv1GS:b-iq8gj0) ([raw file](https://github.com/cockroachdb/cockroach/blob/6963da28b440c20d42bff997917957ea5f628407/pkg/sql/distsql_physical_planner.go#L3905)):*
> ```Go
> // processors eagerly move into the draining state which will cancel the context
> // of parallel TableReaders which might "poison" the transaction.
> func checkScanParallelizationIfLocal(
> ```

I think this needs a unit test


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68679)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2, @RaduBerinde, and @rytaft)

___
*[-- commits, line 36 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-MhnVnV_5cw2o03ClQEk:-Mho40i733aTOiYSsFdX:b7tu2qn) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3dc37a73851322efe7c102e4f66974a52c3c41a/-- commits#L36)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

What else is this setting used for? Will setting it to 0 have any other unintended effects?
</blockquote></details>

Extended the commit message.
___
*[pkg/sql/distsql_physical_planner.go, line 1410 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-MhPnDsnDoJBgqqxlSu7:-Mho45KZ7OEYqinKueHi:bl97fkm) ([raw file](https://github.com/cockroachdb/cockroach/blob/3aa2e824dae305ce536e597f5cd28fe534eea781/pkg/sql/distsql_physical_planner.go#L1410)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I think having a limit is a good idea. Seems useful to prevent a small number of queries from saturating all resources.
</blockquote></details>

Sounds good, let's keep it then.
___
*[pkg/sql/distsql_physical_planner.go, line 3905 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-MhnZseJ7K2CsX8lv1GR:-Mhp8R2RAF2X-djGw33m:bfsf6qq) ([raw file](https://github.com/cockroachdb/cockroach/blob/6963da28b440c20d42bff997917957ea5f628407/pkg/sql/distsql_physical_planner.go#L3905)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I think this needs a unit test
</blockquote></details>

Good point, added. This actually revealed a bug in the indexJoinNode case.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice work!

(Don't forget to update the PR description with the latest state of things)

> Reviewed 2 of 2 files at r23, 3 of 3 files at r24, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-:-MhxJV-22WKjOmAkL5Nz:b-xt042k)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @michae2, @RaduBerinde, @rytaft, and @yuzefovich)

___
*[-- commits, line 79 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-MhxHsaJ0DhNDUwfWqmz):*
nit: do you want to rename this so it's not WIP?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-:-MhyJUONAoEAugedkFgG:b-oduykh)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @michae2, @RaduBerinde, and @rytaft)

___
*[-- commits, line 79 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/68679#-MhxHsaJ0DhNDUwfWqmz):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: do you want to rename this so it's not WIP?
</blockquote></details>

I squashed the commit into the previous one.


<!-- Sent from Reviewable.io -->
